### DISCUSSION TITLE
[RFC]Address validation for country area

### DISCUSSION THREAD

**Discussion by jakubkuc** [saleor/saleor#14443] [RFC]
The proposed solution for #11789 and some other problems with saving country area in addresses.

We want to be able to save country areas in Address objects even if they are not allowed by `i18n` plugin. 
Also, we would like to be able to pass alternative namings for country areas and map them to unify output saved in db.  

### General assumptions
- we will create a mapping for known country areas for different countries
- we will leave the i18n validation as it is now 
- additionally, if `countryArea` is provided in the input it will check the created map if country area provided is valid
- If there are no country area choices in `i18n` for a given country - Saleor will allow passing any string (providing it will pass through our map).

If `i18n` is not allowing country area for a given country it will be saved in db if it is present in our map. 
If `i18n` is allowing country area for given country but it doesn't pass the validation - it will also be put through our map to check if we consider it valid. 

EDIT:
I want to add a check that will add country areas from our map to country area choices for every country address form.

### Notes and issues found 

1.  AddressUpdate mutation does not work optimal. It Overwrites all the fields instead of only given ones. 
e.g : `AddressUpdate` mutation if given fields in input will update those fields with new values and update the ones that are not provided with "Null". ATM it is handled in the dashboard by always sending all fields values. 

2. The second problem (but connected to the first) is old way (connected to templates used in the past) of handling Address objects by using Django forms. We should think about a new way of handling all Address creation/updates without Django forms. This way Saleor would be more flexible during validation and all address operations could be more consistent than now. 

3. We would need a country areas and their allowed alternatives handed to us so we can add them to Saleor core internal mapping mechanism. 

DOD:
- If country area is allowed by the core country area map, user is able to add country area even if country area field is not allowed by `i18n` validation.
- If country area is allowed by core country area map, user is able to add country area even if it's considered invalid by `i18n` validation
- If user is providing all fields according to `i18n` rules everything works the same as before

**Comment by szymi-** [comment]
So if my understanding is correct, if I input a country area that matches what's in Saleor's map of valid country areas, but does not match exactly what is in `i18n`, then the value from `i18n` is saved in Saleor's database. What will I get when I read the value later, e.g. via a graphql query? Will the original value, or value from `i18n`?

**Comment by jakubkuc** [comment]
@szymi- the main point is to keep the output consistent. So in this case it will be what is stored in db. We want to be flexible in terms of entering input, to figure out which country area the user has in mind, without enforcing him to spell it correctly. E.g. like inputting `California` will be saved and fetched as `CA` since that is expected by the local post office.

**Comment by szymi-** [comment]
It would be good have a way to retrieve the originally input value along with what's in the database.
